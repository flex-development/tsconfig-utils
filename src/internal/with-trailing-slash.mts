/**
 * @file Internal - withTrailingSlash
 * @module tsconfig-utils/internal/withTrailingSlash
 */

import pathe from '@flex-development/pathe'

export default withTrailingSlash

/**
 * Ensure `url.href` and `url.pathname` end with a single trailing slash.
 *
 * @internal
 *
 * @param {URL} url
 *  URL to handle
 * @return {URL}
 *  `url`
 */
function withTrailingSlash(url: URL): URL

/**
 * Ensure `str` ends with a single trailing slash.
 *
 * @internal
 *
 * @param {string} str
 *  String to handle
 * @return {string}
 *  `str`
 */
function withTrailingSlash(str: string): string

/**
 * If `value` is a string, ensure `value` ends with a single trailing slash.\
 * Otherwise, ensure `value.href` and `value.pathname` end with a single
 * trailing slash.
 *
 * @internal
 *
 * @param {URL | string} value
 *  URL or string to handle
 * @return {URL | string}
 *  `value`
 */
function withTrailingSlash(value: URL | string): URL | string

/**
 * Ensure `value` ends with a single trailing slash.
 *
 * @internal
 *
 * @param {URL | string} value
 *  URL or string to handle
 * @return {URL | string}
 *  `value`
 */
function withTrailingSlash(value: URL | string): URL | string {
  if (typeof value === 'string') {
    value = value.replace(/[/\\]+$/, '') + pathe.sep
  } else {
    value.href = withTrailingSlash(value.href)
  }

  return value
}
