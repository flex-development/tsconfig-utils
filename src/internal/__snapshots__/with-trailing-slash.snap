// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:internal/withTrailingSlash > should return \`value\` with one trailing slash 1`] = `
{
  "href": file://\${process.cwd()}/,
  "pathname": \${process.cwd()}/,
}
`;
